File: apps/mobile/ios-native/BridgeAPI.swift
import Foundation
struct BridgeReadyProbe: Equatable {
    let ready: Bool
    let errorMessage: String?
}

enum BridgeAPI {
    private static let defaultTimeout: TimeInterval = 10

    static func normalizeEndpoint(_ value: String) -> String {
        var text = value.trimmingCharacters(in: .whitespacesAndNewlines)
        while text.hasSuffix("/") {
            text.removeLast()
        }
        return text
    }

    static func deriveServerHTTPBaseURL(_ endpoint: String) -> String {
        let normalized = normalizeEndpoint(endpoint)
        if normalized.hasPrefix("ws://") {
            return "http://" + normalized.dropFirst(5)
        }
        if normalized.hasPrefix("wss://") {
            return "https://" + normalized.dropFirst(6)
        }
        if normalized.hasPrefix("http://") || normalized.hasPrefix("https://") {
            return normalized
        }
        return "http://" + normalized
    }

    static func buildRemoteShellURL(baseURL: String, theme: String) -> URL? {
        var components = URLComponents(string: normalizeEndpoint(baseURL) + "/ui/index.html")
        components?.queryItems = [URLQueryItem(name: "codexTheme", value: theme)]
        return components?.url
    }

    static func fetchConnectionTarget(baseURL: String, authToken: String?) async throws -> ConnectionTargetResponse {
        let payload = try await requestJSON(
            url: normalizeEndpoint(baseURL) + "/codex-mobile/connect",
            authToken: authToken
        )
        guard let object = payload as? JSONDictionary else {
            throw NativeHostBridgeError.invalidResponse("Bridge connect response was not a JSON object.")
        }
        let connection = object["connection"] as? JSONDictionary
        let auth = object["auth"] as? JSONDictionary
        let recommendedServerEndpoint =
            (connection?["recommendedServerEndpoint"] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .nilIfEmpty ?? normalizeEndpoint(baseURL)
        let localAuthPage = (object["localAuthPage"] as? String)?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .nilIfEmpty
        return ConnectionTargetResponse(
            bridgeID: (connection?["bridgeId"] as? String)?
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .nilIfEmpty,
            recommendedServerEndpoint: recommendedServerEndpoint,
            authMode: (auth?["mode"] as? String) == "device-token" ? "device-token" : "none",
            localAuthPage: localAuthPage
        )
    }

    static func completeDevicePairing(baseURL: String, pairingCode: String, authToken: String?) async throws -> PairingResponse {
        let requestBody: JSONDictionary = [
            "code": pairingCode.trimmingCharacters(in: .whitespacesAndNewlines),
            "deviceName": "Codex Mobile (iOS Native Host)",
        ]
        let payload = try await requestJSON(
            url: normalizeEndpoint(baseURL) + "/auth/pair/complete",
            method: "POST",
            authToken: authToken,
            requestBody: requestBody
        )
        guard let object = payload as? JSONDictionary else {
            throw NativeHostBridgeError.invalidResponse("Pairing response was not a JSON object.")
        }
        let accessToken = (object["accessToken"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        guard !accessToken.isEmpty else {
            throw NativeHostBridgeError.invalidResponse("Pairing response did not include an access token.")
        }
        return PairingResponse(accessToken: accessToken, approved: object["approved"] as? Bool ?? false)
    }

    static func performDirectRPC(baseURL: String, method: String, params: JSONDictionary?, authToken: String?) async throws -> JSONDictionary {
        let requestBody: JSONDictionary = [
            "method": method,
            "params": params ?? [:],
        ]
        let payload = try await requestJSON(
            url: normalizeEndpoint(baseURL) + "/codex-mobile/rpc",
            method: "POST",
            authToken: authToken,
            requestBody: requestBody
        )
        guard let object = payload as? JSONDictionary else {
            throw NativeHostBridgeError.invalidResponse("Direct RPC response was not a JSON object.")
        }
        guard object["ok"] as? Bool == true else {
            let error = (object["error"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines)
            throw NativeHostBridgeError.requestFailed(error?.isEmpty == false ? error! : "Direct RPC failed.")
        }
        return object["result"] as? JSONDictionary ?? [:]
    }

    static func fetchGlobalStateSnapshot(baseURL: String, authToken: String?) async throws -> JSONDictionary {
        var lastError: Error?
        for attempt in 0..<6 {
            do {
                return try await performDirectRPC(
                    baseURL: baseURL,
                    method: "get-global-state-snapshot",
                    params: [:],
                    authToken: authToken
                )
            } catch {
                lastError = error
                if attempt == 5 {
                    throw error
                }
                try await Task.sleep(nanoseconds: 250_000_000)
            }
        }
        throw lastError ?? NativeHostBridgeError.requestFailed("Failed to fetch bridge global state snapshot.")
    }

    static func probeBridgeReady(baseURL: String, authToken: String?) async -> BridgeReadyProbe {
        do {
            _ = try await requestJSON(
                url: normalizeEndpoint(baseURL) + "/readyz",
                authToken: authToken
            )
            return BridgeReadyProbe(ready: true, errorMessage: nil)
        } catch {
            return BridgeReadyProbe(
                ready: false,
                errorMessage: (error as NSError).localizedDescription.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty
            )
        }
    }

    static func proxyHTTPFetch(
        url: URL,
        method: String?,
        requestBody: String?,
        requestHeaders: JSONDictionary?,
        authToken: String?
    ) async throws -> HttpProxyResponse {
        var request = URLRequest(url: url, timeoutInterval: defaultTimeout)
        request.httpMethod = method?.trimmingCharacters(in: .whitespacesAndNewlines).uppercased().nilIfEmpty ?? "GET"
        requestHeaders?.forEach { key, value in
            if let text = value as? String, !text.isEmpty {
                request.setValue(text, forHTTPHeaderField: key)
            }
        }
        if request.value(forHTTPHeaderField: "Authorization").nilIfEmpty == nil,
           let authToken, !authToken.isEmpty {
            request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        }
        if let requestBody, !requestBody.isEmpty {
            request.httpBody = Data(requestBody.utf8)
            if request.value(forHTTPHeaderField: "Content-Type").nilIfEmpty == nil {
                request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            }
        }
        let (data, response) = try await URLSession.shared.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NativeHostBridgeError.invalidResponse("Proxy response did not include HTTP metadata.")
        }
        let headers = httpResponse.allHeaderFields.reduce(into: JSONDictionary()) { partialResult, entry in
            if let key = entry.key as? String {
                partialResult[key] = "\(entry.value)"
            }
        }
        let text = String(data: data, encoding: .utf8) ?? ""
        return HttpProxyResponse(
            body: parseJSONValue(from: text),
            status: httpResponse.statusCode,
            headers: headers
        )
    }

    private static func requestJSON(
        url: String,
        method: String = "GET",
        authToken: String? = nil,
        requestBody: JSONDictionary? = nil
    ) async throws -> Any {
        guard let requestURL = URL(string: url) else {
            throw NativeHostBridgeError.invalidURL("Invalid bridge URL: \(url)")
        }
        var request = URLRequest(url: requestURL, timeoutInterval: defaultTimeout)
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let authToken, !authToken.isEmpty {
            request.setValue("Bearer \(authToken)", forHTTPHeaderField: "Authorization")
        }
        if let requestBody {
            request.httpBody = try JSONSerialization.data(withJSONObject: requestBody, options: [])
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        }

        let configuration = URLSessionConfiguration.default
        configuration.timeoutIntervalForRequest = defaultTimeout
        configuration.timeoutIntervalForResource = defaultTimeout
        let session = URLSession(configuration: configuration)
        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NativeHostBridgeError.invalidResponse("Bridge response did not include HTTP metadata.")
        }
        let responseText = String(data: data, encoding: .utf8) ?? ""
        if !(200...299).contains(httpResponse.statusCode) {
            let message: String
            if let errorObject = parseJSONValue(from: responseText) as? JSONDictionary,
               let errorText = (errorObject["error"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines),
               !errorText.isEmpty {
                message = errorText
            } else if !responseText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
                message = responseText.trimmingCharacters(in: .whitespacesAndNewlines)
            } else {
                message = "Request failed with HTTP \(httpResponse.statusCode)."
            }
            throw NativeHostBridgeError.requestFailed(message)
        }

        if responseText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return [:]
        }
        guard let object = parseJSONValue(from: responseText) else {
            throw NativeHostBridgeError.invalidResponse("Bridge response was not valid JSON.")
        }
        return object
    }
}

File: apps/mobile/android/app/src/main/java/com/boomyao/codexmobile/nativehost/BridgeApi.kt
package com.boomyao.codexmobile.nativehost

import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

object BridgeApi {
    private const val DEFAULT_TIMEOUT_MS = 10_000

    data class BridgeReadyProbe(
        val ready: Boolean,
        val errorMessage: String? = null,
    )

    fun normalizeEndpoint(value: String): String = value.trim().trimEnd('/')

    fun deriveServerHttpBaseUrl(endpoint: String): String {
        val normalized = normalizeEndpoint(endpoint)
        return when {
            normalized.startsWith("ws://") -> "http://${normalized.removePrefix("ws://")}"
            normalized.startsWith("wss://") -> "https://${normalized.removePrefix("wss://")}"
            normalized.startsWith("http://") || normalized.startsWith("https://") -> normalized
            else -> "http://$normalized"
        }
    }

    fun buildRemoteShellUrlFromBaseUrl(baseUrl: String): String {
        return "${normalizeEndpoint(baseUrl)}/ui/index.html"
    }

    fun fetchConnectionTargetByBaseUrl(
        baseUrl: String,
        authToken: String?,
    ): ConnectionTargetResponse {
        val responseText = requestJson(
            url = "${normalizeEndpoint(baseUrl)}/codex-mobile/connect",
            authToken = authToken,
        )
        val payload = JSONObject(responseText)
        val connection = payload.optJSONObject("connection")
        val auth = payload.optJSONObject("auth")
        val recommendedServerEndpoint =
            connection?.optString("recommendedServerEndpoint")?.trim().orEmpty().ifEmpty {
                normalizeEndpoint(baseUrl)
            }
        val localAuthPage = payload.optString("localAuthPage").trim().ifEmpty { null }
        return ConnectionTargetResponse(
            bridgeId = connection?.optString("bridgeId")?.trim()?.ifEmpty { null },
            recommendedServerEndpoint = recommendedServerEndpoint,
            authMode = if (auth?.optString("mode") == "device-token") "device-token" else "none",
            localAuthPage = localAuthPage,
        )
    }

    fun completeDevicePairingByBaseUrl(
        baseUrl: String,
        pairingCode: String,
        authToken: String?,
    ): PairingResponse {
        val body = JSONObject()
            .put("code", pairingCode.trim())
            .put("deviceName", "Codex Mobile (Android Native Host)")
            .toString()
        val responseText = requestJson(
            url = "${normalizeEndpoint(baseUrl)}/auth/pair/complete",
            method = "POST",
            authToken = authToken,
            requestBody = body,
        )
        val payload = JSONObject(responseText)
        val accessToken = payload.optString("accessToken").trim()
        if (accessToken.isEmpty()) {
            throw IllegalStateException("Pairing response did not include an access token.")
        }
        return PairingResponse(
            accessToken = accessToken,
            approved = payload.optBoolean("approved", false),
        )
    }

    fun performDirectRpcByBaseUrl(
        baseUrl: String,
        method: String,
        params: JSONObject?,
        authToken: String?,
    ): JSONObject {
        val body = JSONObject()
            .put("method", method)
            .put("params", params ?: JSONObject())
            .toString()
        val responseText = requestJson(
            url = "${normalizeEndpoint(baseUrl)}/codex-mobile/rpc",
            method = "POST",
            authToken = authToken,
            requestBody = body,
        )
        val payload = JSONObject(responseText)
        if (!payload.optBoolean("ok", false)) {
            throw IllegalStateException(
                payload.optString("error").trim().ifEmpty { "Direct RPC failed." },
            )
        }
        return payload.optJSONObject("result") ?: JSONObject()
    }

    fun fetchGlobalStateSnapshotByBaseUrl(
        baseUrl: String,
        authToken: String?,
    ): JSONObject {
        var lastError: Exception? = null
        repeat(6) { attempt ->
            try {
                return performDirectRpcByBaseUrl(
                    baseUrl = baseUrl,
                    method = "get-global-state-snapshot",
                    params = JSONObject(),
                    authToken = authToken,
                )
            } catch (error: Exception) {
                lastError = error
                if (attempt == 5) {
                    throw error
                }
                Thread.sleep(250L)
            }
        }
        throw lastError ?: IllegalStateException("Failed to fetch bridge global state snapshot.")
    }

    fun isBridgeReadyByBaseUrl(
        baseUrl: String,
        authToken: String?,
    ): Boolean {
        return probeBridgeReadyByBaseUrl(baseUrl, authToken).ready
    }

    fun probeBridgeReadyByBaseUrl(
        baseUrl: String,
        authToken: String?,
    ): BridgeReadyProbe {
        return try {
            requestJson(
                url = "${normalizeEndpoint(baseUrl)}/readyz",
                authToken = authToken,
            )
            BridgeReadyProbe(ready = true)
        } catch (error: Exception) {
            BridgeReadyProbe(
                ready = false,
                errorMessage = error.message?.trim().orEmpty().ifEmpty { null },
            )
        }
    }

    private fun requestJson(
        url: String,
        method: String = "GET",
        authToken: String? = null,
        requestBody: String? = null,
    ): String {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.connectTimeout = DEFAULT_TIMEOUT_MS
        connection.readTimeout = DEFAULT_TIMEOUT_MS
        connection.requestMethod = method
        connection.setRequestProperty("Accept", "application/json")
        if (!authToken.isNullOrBlank()) {
            connection.setRequestProperty("Authorization", "Bearer ${authToken.trim()}")
        }
        if (requestBody != null) {
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.outputStream.use { stream ->
                stream.write(requestBody.toByteArray())
            }
        }

        val responseCode = connection.responseCode
        val stream = if (responseCode in 200..299) {
            connection.inputStream
        } else {
            connection.errorStream ?: connection.inputStream
        }
        val text = BufferedReader(InputStreamReader(stream)).use { reader ->
            buildString {
                while (true) {
                    val line = reader.readLine() ?: break
                    append(line)
                }
            }
        }
        if (responseCode !in 200..299) {
            val message = runCatching {
                JSONObject(text).optString("error").trim()
            }.getOrNull().orEmpty()
            if (message.isNotEmpty()) {
                throw IllegalStateException(message)
            }
            val fallbackText = text.trim()
            if (fallbackText.isNotEmpty()) {
                throw IllegalStateException(fallbackText)
            }
            throw IllegalStateException("Request failed with HTTP $responseCode.")
        }
        return text
    }
}
